### DISCUSSION TITLE
Can Pytermgui meet my project needs?

### DISCUSSION THREAD

**Discussion by jeffwright13** [bczsalba/pytermgui#51] [Q&A] [answered]
Hello,

I have written a [pytest plugin](https://github.com/jeffwright13/pytest-tui) that launches a TUI at the end of a pytest run, and presents the categorized results to the user. I am wondering if Pytestgui would be able to meet my needs to implement the TUI as an alternative to what I have in place already (one uses Textual, but it has performance issues; the other uses PyTermTk but it also is not perfect, and I have a hard time figuring out how to fix it without the documentation).

I would like to keep the look and feel that I have in the other two TUIs, as it seems to meet my needs, if not the needs of all users (who as far as I can tell are zero at this point :-) ).

Requirements I can think of:
- a Tab (or Notebook) widget
- ANSI-coded text support
- mouse support (clicking/moving/scrolling)
- copy/paste support (nice to have)
- hotkey support, or even keymaps (for those users who do not like using a mouse in a terminal/text environment)
- clean exit from TUI back to terminal

Do you think Pytermgui could fit the bill? Thanks in advance!

**Comment by bczsalba** [comment, marked as answer]
Hey there!

I think most things should be doable, yes. I've actually been watching your project for a bit now, looks very interesting!

I think the exact layout of the TUI probably can't be the exact same for all of these libraries. PyTermTK is very tab-based, and Textual has support for it built in as well. PyTermGUI at the moment doesn't have a tab layout, but implementing one using some buttons or something functionally similar with the upcoming `Collapsible` widget shouldn't be too hard.

All widgets at the end of their run print using ANSI text, so that part should be fine too.

I think it should be good for most, if not all, of your "requirements", and I'd love to see you try to implement it in PTG! Leaving issues about the problems you run into is probably the most efficient way to further improve the library.

Thank you for your interest!

**Reply by jeffwright13** [reply]
I will give it a shot then. I had to write my own Tab/Tabs widget in Textual because the main code branch doesn't have that widget (although they have a real slick one in their css branch, it's not yet ready for public consumption).

I'll see what I can figure out as far as implementing some sort of tabbed widget in PyTermGui. Expect some more questions! :-)
